### DISCUSSION TITLE
How to release the memory of vsg::Array？

### DISCUSSION THREAD

**Discussion by jacky-wang7** [vsg-dev/VulkanSceneGraph#1276] [Q&A] [answered]
The function I tested only has these codes,Here is my code：
```
void test()
{
  //Big data for easy observation.
  auto arr = vsg::vec4Array::create(1000000,vsg::vec4(1,1,1,1));
  return;
}
```
I thought that at the end of the function, the smart pointer would be destroyed and the memory would be released. However, from the task manager, the memory was not released.
So I tried to release it manually, but it still didn't work.
```
void test()
{
  //Big data for easy observation.
  auto arr = vsg::vec4Array::create(1000000,vsg::vec4(1,1,1,1));
  arr->clear();
  return;
}
```

**Comment by robertosfield** [comment, marked as answer]
The VSG uses the [vsg::Allocator](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/core/Allocator.h) class to do block allocation with affinity between allocation types so similar object/data types get allocated in the same blocks of memory.  This is done to improve runtime performance and can make a big difference for applications that are bandwidth limited - real-time graphics is huge bandwidth hog even on the CPU so it's crucial for getting best performance.

I recently improved the internal implementation to speed up allocations/deallocations and wrote up this work on this forum. This thread should also give you an idea what it's all about along with performance metrics: https://github.com/vsg-dev/VulkanSceneGraph/discussions/1228

So in your case the vsg::Array and it's allocated data will be allocated in a block of memory that the VSG will keep around until either those blocks are released or at the end of the life of the application.  If you want to release any empty blocks just call:

~~~ cpp
    vsg::Allocator::instance()->deleteEmptyMemoryBlocks();
~~~

**Comment by jacky-wang7** [comment]
I get it, I'm going to study this, there's more involved than I expected. Thank you so much.
